Enable automatic URL linking (bug 1019475)

ryzokuken opened the pull request:
Automatically detect links in the text content of a file and automatically generate link annotations at the appropriate locations to achieve automatic link detection and hyperlinking.

References:
* https://bugzilla.mozilla.org/show_bug.cgi?id=1019475
* https://github.com/mozilla/pdf.js/issues/3172

ryzokuken committed:
Enable automatic URL linking

Automatically detect links in the text content of a file and automatically
generate link annotations at the appropriate locations to achieve
automatic link detection and hyperlinking.

Snuffleupagus reviewed (changes requested):
How does this perform, especially in documents that contain a lot of text?

Also, we probably want a new option/preference to be able to disable this functionality.

ryzokuken commented on web/annotation_layer_builder.js:
```diff
@@ -119,6 +119,8 @@ class AnnotationLayerBuilder {
     }
 
+    annotations.push(...linkAnnotations);
+
```
Great catch, I disabled injecting link annotations if _any_ annotations exist atm, with a note that we should switch to something smarter for a later iteration.

ryzokuken commented on web/pdf_page_view.js:
```diff
@@ -1197,6 +1204,79 @@ class PDFPageView {
+          style: 1,
+          dashArray: [3],
+          horizontalCornerRadius: 0,
+          verticalCornerRadius: 0
```
Sorry I missed that, I was fiddling around with these values and forgot to reset them. Ideally we shouldn't be setting them by hand so I added a note to replace this whole thing with something like a `new LinkAnnotation` call.

ryzokuken commented on web/pdf_page_view.js:
```diff
@@ -1197,6 +1204,79 @@ class PDFPageView {
+        isEditable: false,
+        hasApperance: false,
+        modificationDate: null,
+        structParent: 2,
```
Apparently this properly wasn't needed to make things work so I removed it alongside other redundant properties.

ryzokuken commented on web/pdf_page_view.js:
```diff
@@ -1197,6 +1204,79 @@ class PDFPageView {
+        rect,
+        // NOTE boilerplate-y
+        annotationType: 2,
+        annotationFlags: 4,
```
Removed `annotationFlags` and the only remaining hardcoded number now is `annotationType`. Ideally I should still not be assigning any of these manually and using a constructor.

ryzokuken replied:
No more hard-coded numbers.

ryzokuken commented on web/pdf_page_view.js:
```diff
@@ -1197,6 +1204,79 @@ class PDFPageView {
+  #processLinks() {
+    return this.pdfPage.getTextContent().then(content => {
+      const [text, diffs] = normalizedTextContent(content);
+      const urlRegex = /\b(?:https?:\/\/|mailto:|www.)(?:[[\S--\[]--\p{P}]|\/|[\p{P}--\[]+[[\S--\[]--\p{P}])+/gmv;
```
This might no longer be relevant since it's now a static field on the class.

ryzokuken commented:
@Snuffleupagus

>  Besides, it isn't necessary since the textContent is already available once the textLayer has rendered; see

I was a bit unsure if I understood exactly what you were suggesting but how does this commit look? It "fetches" the textContents from the previous render of the textLayer and makes the processing step sync.

https://github.com/mozilla/pdf.js/pull/19110/commits/93e5417a4e1dfcdbe7a2a9adab9ae7d917445681

ryzokuken commented:
@Snuffleupagus nvm my last comment, I figured it out after looking at `pdfPageView._textHighlighter.textDivs` a couple of times it occurred to me what you were talking about.

https://github.com/mozilla/pdf.js/pull/19110/files#diff-71772f56be799df522c2076ab5fa476253ef15c607af5812536c41696d97cd59R73

calixteman commented on web/annotation_layer_builder.js:
```diff
@@ -119,6 +119,20 @@ class AnnotationLayerBuilder {
+        if (
+          annotation.subtype === "Link" &&
+          annotation.url === link.url &&
+          Util.intersect(annotation.rect, link.rect) !== null
```
In order to avoid some corner case bug, maybe a better way to do would be to compute the ratio area(intersection) / area(annotation.rect) and if it's greater than a threshold then consider that links are very likely the same.

nicolo-ribaudo commented on web/autolinker.js:
```diff
@@ -0,0 +1,125 @@
+    /\b(?:https?:\/\/|mailto:|www\.)(?:[[\S--\[]--\p{P}]|\/|[\p{P}--\[]+[[\S--\[]--\p{P}])+|\b[\S--@]+@[\S--.]+.[[\S--\[]--\p{P}]{2,}/gmv; // Regex can be tested and verified at https://regex101.com/r/zuMeQX/3.
+
+  static #addLinkAnnotations(url, index, length, pdfPageView) {
+    const convertedMatch = pdfPageView._textHighlighter._convertMatches(
```
If we are using `_convertMatches` outside of the `_textHighlighter` class, we should probably remove the `_` from its name.

ryzokuken replied:
What about `_textHighlighter` itself?

ryzokuken replied:
No one seconded this comment, so I'm keeping things as-is for now.

calixteman commented on test/unit/autolinker_spec.js:
```diff
@@ -0,0 +1,166 @@
+    const matches = Autolinker.findLinks(emails.join("\n"));
+    expect(matches.length).toEqual(12);
+    expect(matches[0].url).toEqual("mailto:username@example.com");
+    expect(matches[1].url).toEqual("mailto:someone@subdomain.example.com");
```
You could have an array `[[string, expected]...]` and just loop on it.

calixteman commented on web/autolinker.js:
```diff
@@ -0,0 +1,125 @@
+
+    const pageBox = pdfPageView.textLayer.div.getBoundingClientRect();
+    const linkAnnotations = [];
+    for (const linkBox of range.getClientRects()) {
```
nit: `{ left, top, ... }` instead of `linkBox`.

Snuffleupagus replied:
Sorry, but generating separate annotations for multiline links seems wrong and it'll easily lead to duplicates not being found as intended.

The *correct* way to handle multiline links is to generate `quadPoints` for those, note e.g. https://github.com/mozilla/pdf.js/blob/711bf2bd1251a247881fb14d6c4185214eb36e16/src/core/annotation.js#L3785-L3788

calixteman commented on web/autolinker.js:
```diff
@@ -0,0 +1,125 @@
+    const [normalizedText, diffs] = normalize(text);
+    const matches = normalizedText.matchAll(Autolinker.#urlRegex);
+    const links = [];
+    for (const match of matches) {
```
nit: `const [matched] = match`.

ryzokuken replied:
We check `match.index` for finding the original index of the text.

calixteman commented on web/autolinker.js:
```diff
@@ -0,0 +1,125 @@
+    const links = [];
+    for (const match of matches) {
+      const raw =
+        match[0].startsWith("www.") ||
```
nit: we just need to check for the first char, right ?

ryzokuken replied:
No, for emails that start with a `w` that would not work.

calixteman replied:
Ok, so just in case someone has the same idea, we must have a test for that.

ryzokuken replied:
I was the one who had that idea! (for http)

Added tests and fixed the bug for emails that started with "http".

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -97,7 +97,7 @@ class AnnotationLayerBuilder {
    *   annotations is complete.
    */
-  async render(viewport, options, intent = "display") {
+  async render(viewport, options, intent = "display", linkAnnotations) {
```
The new parameter is not documented, also the name should be clearer in that these are not "normal" LinkAnnotations; e.g.
```suggestion
  async render(viewport, options, intent = "display", inferredLinkAnnotations) {
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -119,6 +119,24 @@ class AnnotationLayerBuilder {
+          Math.abs(rect[2] - rect[0]) * Math.abs(rect[3] - rect[1]);
+        const intersect = Util.intersect(annotation.rect, link.rect); // Find the intersection between the annotation and the link.
+        if (
+          annotation.subtype === "Link" &&
```
This should compare against the `annotationType`, rather than using string comparison.

Snuffleupagus commented on web/app.js:
```diff
@@ -493,6 +493,7 @@ const PDFViewerApplication = {
       abortSignal: this._globalAbortController.signal,
       enableHWA,
       supportsPinchToZoom: this.supportsPinchToZoom,
+      enableAutolinking: AppOptions.get("enableAutolinking"),
```
Here, and elsewhere throughout the patch.
```suggestion
      enableAutoLinking: AppOptions.get("enableAutoLinking"),
```

Snuffleupagus commented on web/app_options.js:
```diff
@@ -236,6 +236,12 @@ const defaultOptions = {
     value: false,
     kind: OptionKind.VIEWER + OptionKind.PREFERENCE,
   },
+  enableAutolinking: {
+    // TODO: remove it when unnecessary.
+    /** @type {boolean} */
+    value: false,
+    kind: OptionKind.VIEWER,
+  },
```
- This isn't sorted alphabetically (and please fix the casing).
 - I don't think this should be temporary, since even Adobe Reader allows this to be configured by the user. (Personally I really want to be able to disable this feature.)
 - It should be a preference as well.

```suggestion
  enableAutoLinking: {
    /** @type {boolean} */
    value: false,
    kind: OptionKind.VIEWER + OptionKind.PREFERENCE,
  },
```

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -0,0 +1,125 @@
+
+class Autolinker {
+  static #urlRegex =
+    /\b(?:https?:\/\/|mailto:|www\.)(?:[[\S--\[]--\p{P}]|\/|[\p{P}--\[]+[[\S--\[]--\p{P}])+|\b[[\S--@]--\{]+@[\S--.]+\.[[\S--\[]--\p{P}]{2,}/gmv; // Regex can be tested and verified at https://regex101.com/r/zgDwPE/1.
```
The above error *must* be addressed.

Also, as I've previously mentioned, this regular expression should be initialized *lazily*.

ryzokuken replied:
The regexp is now initialized lazily.

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -0,0 +1,125 @@
+        unsafeUrl: url,
+        url,
+        rect,
+        annotationType: 2,
```
Again, as previously mentioned, don't hard-code this value; note https://github.com/mozilla/pdf.js/blob/711bf2bd1251a247881fb14d6c4185214eb36e16/src/shared/util.js#L133

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -148,6 +151,8 @@ class PDFPageView {
     regularAnnotations: true,
   };
 
+  #linkAnnotations = [];
```
As mentioned before, I'd suggest e.g. `this.#inferredLinkAnnotations` instead.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -177,6 +182,7 @@ class PDFPageView {
       options.maxCanvasPixels ?? AppOptions.get("maxCanvasPixels");
     this.pageColors = options.pageColors || null;
     this.#enableHWA = options.enableHWA || false;
+    this.#enableAutolinking = options.enableAutolinking || false;
```
This parameter isn't documented.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1090,9 +1097,13 @@ class PDFPageView {
         );
 
-        this.#renderTextLayer();
+        const textLayerP = this.#renderTextLayer();
```
No single letter abbreviations please.
```suggestion
        const textLayerPromise = this.#renderTextLayer();
```

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -321,6 +323,7 @@ class PDFViewer {
     this.#mlManager = options.mlManager || null;
     this.#enableHWA = options.enableHWA || false;
     this.#supportsPinchToZoom = options.supportsPinchToZoom !== false;
+    this.#enableAutolinking = options.enableAutolinking || false;
```
This parameter isn't documented.

Snuffleupagus commented on test/integration/autolinker_spec.mjs:
```diff
@@ -0,0 +1,75 @@
+      await closePages(pages);
+    });
+
+    it("must not add links when not necessary", async () => {
```
Nit:
```suggestion
    it("must not add links when unnecessary", async () => {
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -119,6 +119,24 @@ class AnnotationLayerBuilder {
 
+    const uniqueLinks = linkAnnotations.filter(link => {
+      for (const annotation of annotations) {
+        const area = rect =>
+          Math.abs(rect[2] - rect[0]) * Math.abs(rect[3] - rect[1]);
+        const intersect = Util.intersect(annotation.rect, link.rect); // Find the intersection between the annotation and the link.
```
This must also take `quadPoints` into account, when those exist, in addition to just the `rect`.

ryzokuken replied:
This one turned out to be a bit tricky, also I was a bit unsure how to deal with comparing annotations with `quadPoints` and those without. It was also harder than I expected to find decent examples of multi-line links in PDFs in the wild but check out what I did and let me know how it looks.

Snuffleupagus replied:
> I tried myself to build an example in either Word or with LaTeX and I didn't manage to have one: a splited url gives different annotations for each chunk.
> So it's likely possible to craft something (and I don't know how it'd supported in the different viewers) but I'm not sure it's worth spending too much time and that. So we could just add a TODO and wait for a bug.

@calixteman So you're suggesting that all quadPoints-handling can be removed from the patch (with a TODO comment about that)? If so, I suppose that's fine for now.

ryzokuken replied:
Marking this as addressed since the lack of a test case is in a TODO comment now.

Snuffleupagus replied:
This isn't fully resolved, since it's not clear (to me) if the suggestion was that we should remove all quadPoints handling or not!?

calixteman replied:
For me it's fine to keep the intersection computation using quadPoints even if we don't have a test case.
And at some point, if we've a test case demonstrating that this code is wrong, then we'll fix and will add a test for it.

Snuffleupagus reviewed (changes requested):
As asked already in https://github.com/mozilla/pdf.js/pull/19110#pullrequestreview-2462265413, how does all of this extra parsing affect performance?

(Especially since it'll delay rendering of both the AnnotationLayer *and* the AnnotationEditorLayer.)

ryzokuken commented on web/autolinker.js:
```diff
@@ -0,0 +1,149 @@
+  static findLinks(text) {
+    const [normalizedText, diffs] = normalize(text);
+    const matches = normalizedText.matchAll(
+      /\b(?:https?:\/\/|mailto:|www\.)(?:[[\S--\[]--\p{P}]|\/|[\p{P}--\[]+[[\S--\[]--\p{P}])+|\b[[\S--@]--\{]+@[\S--.]+\.[[\S--\[]--\p{P}]{2,}/gmv // Regex can be tested and verified at https://regex101.com/r/zgDwPE/1.
```
@Snuffleupagus @calixteman the regexp is now lazily initialized, but the scanner still fails to understand the `v` flag perhaps because it's a relatively new one (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicodeSets).

ryzokuken replied:
https://caniuse.com/mdn-javascript_builtins_regexp_unicodesets

Snuffleupagus replied:
> the regexp is now lazily initialized

You're now initializing it every time it's being used, which seems wasteful and isn't really what I asked for.
Instead you should initialize it *only once*, but do so lazily; e.g. something like the following.
```javascript
class Autolinker {
  static #regex;

  static findLinks(text) {
    // Regex can be tested and verified at https://regex101.com/r/zgDwPE/1.
    this.#regex ??= /\b(?:https?:\/\/|mailto:|www\.)(?:[[\S--\[]--\p{P}]|\/|[\p{P}--\[]+[[\S--\[]--\p{P}])+|\b[[\S--@]--\{]+@[\S--.]+\.[[\S--\[]--\p{P}]{2,}/gmv

    const [normalizedText, diffs] = normalize(text);
    const matches = normalizedText.matchAll(this.#regex);
    ...
```

Snuffleupagus replied:
> but the scanner still fails to understand the `v` flag perhaps because it's a relatively new one (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicodeSets).

That's potentially problematic then, note https://github.com/mozilla/pdf.js/wiki/Frequently-Asked-Questions#faq-support

Snuffleupagus replied:
> and Babel supports transpiling it when in the legacy build :)

Does it do that by default? Because if not, this PR obviously needs to include that functionality as well.

ryzokuken replied:
@Snuffleupagus fixed the issue with lazy initialization.

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -0,0 +1,149 @@
+    annotationType: AnnotationType.LINK,
+    rotation: 0,
+    ...calculateLinkPosition(range, pdfPageView),
+    // This is just the default for AnnotationBorderStyle. At some point we
+    // should switch to something better like `new LinkAnnotation` here.
```
Given that `new LinkAnnotation` is *worker-thread functionality* there's simply no good/efficient way to use that here, so let's drop that part of the comment.
```suggestion
    // This is just the default for AnnotationBorderStyle.
```

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -0,0 +1,149 @@
+ * limitations under the License.
+ */
+
+import {
+  AnnotationBorderStyleType,
+  AnnotationType,
+  createValidAbsoluteUrl,
+  Util,
+} from "../src/shared/util.js";
```
https://github.com/mozilla/pdf.js/pull/18900#discussion_r1809406557 applies here as well.

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -25,7 +25,8 @@
 
-import { AnnotationLayer } from "pdfjs-lib";
+import { AnnotationLayer, Util } from "pdfjs-lib";
+import { AnnotationType } from "../src/shared/util.js";
```
https://github.com/mozilla/pdf.js/pull/18900#discussion_r1809406557 applies here as well.

ryzokuken replied:
Sorry, I got tricked by a tool from my workflow automatically adding these.

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -128,6 +128,7 @@ function isValidAnnotationEditorMode(mode) {
  *   rendering. The default value is `false`.
  * @property {boolean} [supportsPinchToZoom] - Enable zooming on pinch gesture.
  *   The default value is `true`.
+ * @property {boolean} [enableAutoLinking] - Enable automatic linking of URLs.
```
This is perhaps not all that helpful if you don't already know what the feature does.
```suggestion
 * @property {boolean} [enableAutoLinking] - Enable creation of hyperlinks from
     text that look like URLs. The default value is `false`.
```

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -84,6 +85,7 @@ import { XfaLayerBuilder } from "./xfa_layer_builder.js";
  *   the necessary layer-properties.
  * @property {boolean} [enableHWA] - Enables hardware acceleration for
  *   rendering. The default value is `false`.
+ * @property {boolean} [enableAutoLinking] - Enable automatic linking of URLs.
```
This is perhaps not all that helpful if you don't already know what the feature does.
```suggestion
 * @property {boolean} [enableAutoLinking] - Enable creation of hyperlinks from
     text that look like URLs. The default value is `false`.
```

Snuffleupagus commented on test/pdfs/.gitignore:
```diff
@@ -696,3 +696,5 @@
 !issue18911.pdf
 !issue19207.pdf
 !issue19239.pdf
+!link.pdf
+!mozilla.org.pdf
```
Can we please have better named test files, e.g. as follows?
```suggestion
!bug1019475_1.pdf
!bug1019475_2.pdf
```

Snuffleupagus commented on test/unit/autolinker_spec.js:
```diff
@@ -0,0 +1,220 @@
+      ["www.a.com/#a=@?q=rr&r=y", "http://www.a.com/#a=@?q=rr&r=y"],
+      ["http://a.com/1/2/3/4\\5\\6", "http://a.com/1/2/3/4/5/6"],
+      ["http://www.example.com/foo;bar", "http://www.example.com/foo;bar"],
+      // ["www.abc.com/#%%^&&*(", "http://www.abc.com/#%%^&&*("],
```
Here, and elsewhere below, please remove such lines rather than leaving commented-out code around.

ryzokuken replied:
They are edge cases from the PDFium test suite that we don't yet work with. They could be made to work at a later point.

ryzokuken replied:
The test file has been streamlined but the commented out tests are still in there. I don't mind keeping them around personally since they are reminders about the edge cases we could work on but if you feel strongly about this I'll go back and remove them.

calixteman replied:
You could just add something like `TODO: handle this case.` or something like that. An other possibility is to remove them as asked by @Snuffleupagus and once the patch is merged, file a bug with the different cases to be implemented.

@Snuffleupagus would it be fine for you ?

Snuffleupagus replied:
As long as TODO comments are added, I suppose that fine since those can be found easily.

Snuffleupagus replied:
This isn't addressed in the latest version.

ryzokuken replied:
Addressed it.

calixteman commented on test/unit/autolinker_spec.js:
```diff
@@ -0,0 +1,220 @@
+      ],
+    ];
+    const matches = Autolinker.findLinks(links.map(link => link[0]).join("\n"));
+    expect(matches.length).toEqual(3);
```
nit: here and elsewhere in the tests, could you use a loop instead of unrolling it ?

ryzokuken replied:
done!

calixteman commented:
> How does this perform, especially in documents that contain a lot of text?
> 
> Also, we probably want a new option/preference to be able to disable this functionality.

Yep it's better to have a pref on the firefox side too.

That said, I rendered the first 403 pages of `pdf.pdf` (in wrapped mode and with a zoom set to 10) and here's a profile:
https://share.firefox.dev/42mf3uA

and the overall time spent in `processLinks` is ~131ms !! so it shouldn't be a problem.

calixteman commented on web/annotation_layer_builder.js:
```diff
@@ -119,6 +120,61 @@ class AnnotationLayerBuilder {
 
+    const uniqueLinks = inferredLinkAnnotations.filter(link => {
+      for (const annotation of annotations) {
+        const annotationRects = annot => {
```
nit: `({ quadPoints })` and then you can remove `annot.`.

ryzokuken replied:
Sorry there was an error in my code, the non-quadPoints `return` actually should have returned `[annot.rect]`, fixed it now.

calixteman commented on web/annotation_layer_builder.js:
```diff
@@ -119,6 +120,61 @@ class AnnotationLayerBuilder {
+          let totalArea = 0;
+          for (const rect of rects) {
+            totalArea +=
+              Math.abs(rect[2] - rect[0]) * Math.abs(rect[3] - rect[1]);
```
nit: take the overall abs value, there's no need to call it two times.

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -119,6 +120,61 @@ class AnnotationLayerBuilder {
       return;
     }
 
+    const uniqueLinks = inferredLinkAnnotations.filter(link => {
```
```suggestion
    const inferredLinks = inferredLinkAnnotations.filter(link => {
```

Snuffleupagus replied:
Given that the viewer components must still work, this code cannot assume that `inferredLinkAnnotations` is provided.

Also, having all of this code in the middle of `AnnotationLayerBuilder.prototype.render` seems unfortunate and I'd suggest placing the new helper code in a *private method* in this class.
That'd allow this to be written as follows instead:
```suggestion
    if (inferredLinkAnnotations?.length) {
      annotations.push(...this.#parseInferredLinks(inferredLinkAnnotations);
    }
```

Snuffleupagus commented on test/integration/autolinker_spec.mjs:
```diff
@@ -0,0 +1,75 @@
+    let pages;
+
+    beforeAll(async () => {
+      pages = await loadAndWait("mozilla.org.pdf", ".annotationLayer");
```
This file-name no longer exists.

Snuffleupagus commented on test/integration/autolinker_spec.mjs:
```diff
@@ -0,0 +1,75 @@
+    let pages;
+
+    beforeAll(async () => {
+      pages = await loadAndWait("link.pdf", ".annotationLayer");
```
This file-name no longer exists.

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -148,6 +153,8 @@ class PDFPageView {
     regularAnnotations: true,
   };
 
+  #inferredLinkAnnotations = [];
```
```suggestion
  #inferredLinkAnnotations = null;
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -94,10 +94,11 @@ class AnnotationLayerBuilder {
    * @param {PageViewport} viewport
    * @param {Object} options
    * @param {string} intent (default value is 'display')
+   * @param {Array<Object>} inferredLinkAnnotations
```
This must be marked as optional.
```suggestion
   * @param {Array<Object>} [inferredLinkAnnotations]
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -94,10 +94,11 @@ class AnnotationLayerBuilder {
    *   annotations is complete.
    */
-  async render(viewport, options, intent = "display") {
+  async render(viewport, options, intent = "display", inferredLinkAnnotations) {
```
```suggestion
  async render(viewport, options, intent = "display", inferredLinkAnnotations = null) {
```

ryzokuken commented:
@calixteman @Snuffleupagus addressed your suggestions. Thanks for the reviews! The failure at the moment comes from additional objects being exposed from `pdfjs-lib` (I hope that's okay).

Also I updated the tests but for some reason couldn't ensure that the `enableAutoLinking` app option was set properly so let me know if I'm invoking that incorrectly somehow.

ryzokuken commented on web/annotation_layer_builder.js:
```diff
@@ -210,6 +222,61 @@ class AnnotationLayerBuilder {
+          }
+          return [annot.rect];
+        };
+
+        const intersectAnnotations = (annot1, annot2) => {
+          const intersections = [];
+          const annot1Rects = annotationRects(annot1);
+          const annot2Rects = annotationRects(annot2);
+          for (const rect1 of annot1Rects) {
+            for (const rect2 of annot2Rects) {
+              const intersection = Util.intersect(rect1, rect2);
+              if (intersection) {
+                intersections.push(intersection);
+              }
+            }
+          }
+          return intersections;
+        };
+
+        const areaRects = rects => {
+          let totalArea = 0;
+          for (const rect of rects) {
+            totalArea += Math.abs((rect[2] - rect[0]) * (rect[3] - rect[1]));
+          }
+          return totalArea;
+        };
+
+        let intersections; // TODO: Add a test case to verify that we can find the intersection between two annotations with quadPoints properly.
+        if (
+          annotation.annotationType === AnnotationType.LINK &&
+          annotation.url === link.url &&
+          (intersections = intersectAnnotations(annotation, link)).length > 0 &&
+          areaRects(intersections) / areaRects(annotationRects(link)) > 0.5 // If the overlap is more than 50%.
+        ) {
+          return false;
+        }
+      }
+      return true;
+    });
+  }
```
Thanks for the suggestion, done.

ryzokuken commented:
I believe I have addressed every single comment that was still unresolved and have proceeded to mark them as resolved. Please check out the PR now and let me know what you think or if there's something that's still off by any chance.

Snuffleupagus commented:
> and the overall time spent in `processLinks` is ~131ms !! so it shouldn't be a problem.

That's good, however there's one additional side to the story here:
 - Currently, in `master`, the textLayer and annotationLayer will render in parallel.
 - With this patch the annotationLayer will now wait for the textLayer to finish first, and it'll thus take longer than before until the annotationLayer (and thus indirectly the annotationEditorLayer) is rendered.

ryzokuken commented:
@Snuffleupagus that's true, although they would continue to load in parallel if you haven't enabled this flag.

ryzokuken commented:
@calixteman thanks for the pointers! @Snuffleupagus your concern should be addressed now, new profile at https://share.firefox.dev/42yU2NG.

calixteman commented:
/botio test

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -128,24 +135,13 @@ class AnnotationLayerBuilder {
 
     // Create an annotation layer div and render the annotations
     // if there is at least one annotation.
-    const div = (this.div = document.createElement("div"));
-    div.className = "annotationLayer";
-    this.#onAppend?.(div);
```
Why is this removed here, since that seems unrelated!?
Please move this back.

ryzokuken replied:
Done. Also I've set `this.div.hidden` to `false` at the end of `injectLinkAnnotations` as a result, let me know if you prefer a `show` method.

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -177,6 +175,22 @@ class AnnotationLayerBuilder {
   }
 
+  #initAnnotationLayer(viewport, structTreeLayer) {
+    const div = (this.div = document.createElement("div"));
+    div.className = "annotationLayer";
+    this.#onAppend?.(div);
```
*Note the previous comment.*

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -177,6 +175,22 @@ class AnnotationLayerBuilder {
+      annotationEditorUIManager: this._annotationEditorUIManager,
+      page: this.pdfPage,
+      viewport: viewport.clone({ dontFlip: true }),
+      structTreeLayer: structTreeLayer || null,
```
```suggestion
      structTreeLayer,
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -195,6 +209,23 @@ class AnnotationLayerBuilder {
     return !!this.annotationLayer?.hasEditableAnnotations();
   }
 
+  async injectLinkAnnotations(inferredLinks, viewport, structTreeLayer) {
```
- This should take a parameter object (and it needs a JSDoc-typedef as well), similar to how various `render`-methods were recently updated.

 - Also, this must *obviously* check that the annotationLayer is still active.
```suggestion
  async injectLinkAnnotations({ inferredLinks, viewport, structTreeLayer = null }) {
    if (this._cancelled) {
      return;
    }
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -195,6 +209,23 @@ class AnnotationLayerBuilder {
   }
 
+  async injectLinkAnnotations(inferredLinks, viewport, structTreeLayer) {
+    const uniqueLinks = this.#annotations
```
This is a somewhat strange/confusing name, since there's nothing "unique" about these links.
```suggestion
    const newLinks = this.#annotations
```

ryzokuken replied:
Okay

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -0,0 +1,156 @@
+}
+
+function createLinkAnnotation({ url, index, length }, pdfPageView, id) {
+  const convertedMatch = pdfPageView._textHighlighter._convertMatches(
+    [index],
+    [length]
+  )[0];
+
+  const range = new Range();
+  range.setStart(
+    pdfPageView._textHighlighter.textDivs[convertedMatch.begin.divIdx]
+      .firstChild,
+    convertedMatch.begin.offset
+  );
+  range.setEnd(
+    pdfPageView._textHighlighter.textDivs[convertedMatch.end.divIdx].firstChild,
+    convertedMatch.end.offset
+  );
```
Can't this be shortened/simplified a little bit?
```suggestion
  const { _convertMatches, textDivs } = pdfPageView._textHighlighter;
  const [convertedMatch] = _convertMatches([index], [length]);

  const range = new Range();
  range.setStart(
    textDivs[convertedMatch.begin.divIdx].firstChild,
    convertedMatch.begin.offset
  );
  range.setEnd(
    textDivs[convertedMatch.end.divIdx].firstChild,
    convertedMatch.end.offset
  );
```

ryzokuken replied:
Doing this didn't work because `_convertMatches` needs `this` and binding wouldn't have made the line much shorter. I've improved this part of the code in another way, let me know how it looks.

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -195,6 +215,37 @@ class AnnotationLayerBuilder {
+    inferredLinks,
+    viewport,
+    structTreeLayer = null,
+  }) {
```
We should also ensure that `render` has completed *before* this method is invoked, which should be achievable the following way:

 - Apply the following diff.
```diff
diff --git a/web/annotation_layer_builder.js b/web/annotation_layer_builder.js
index b6e31404c..90ae789b3 100644
--- a/web/annotation_layer_builder.js
+++ b/web/annotation_layer_builder.js
@@ -133,6 +133,8 @@ class AnnotationLayerBuilder {
     this.#onAppend?.(div);

     if (annotations.length === 0) {
+      this.#annotations = annotations;
+
       this.hide();
       return;
     }
```

 - Add the following check in this method.
```suggestion
  }) {
    if (this.#annotations === null) {
      throw new Error("`render` must have been called.");
    }
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -195,6 +215,37 @@ class AnnotationLayerBuilder {
+      return;
+    }
+
+    const newLinks = this.#annotations
```
*Given the previous comment.*
```suggestion
    const newLinks = this.#annotations.length
```

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -195,6 +215,37 @@ class AnnotationLayerBuilder {
+    }
+
+    await this.annotationLayer.addLinkAnnotations(newLinks, this.linkService);
+    this.div.hidden = false;
```
This could very easily overwrite a previous `hide`-call from a `PDFPageView`-instance, which could potentially lead to intermittent problems.
The following might be one way to address that:

 - Apply this diff.
```diff
diff --git a/web/annotation_layer_builder.js b/web/annotation_layer_builder.js
index b6e31404c..a8bd11fb2 100644
--- a/web/annotation_layer_builder.js
+++ b/web/annotation_layer_builder.js
@@ -57,6 +57,8 @@ import { PresentationModeState } from "./ui_utils.js";
  */

 class AnnotationLayerBuilder {
+  #externalHide = false;
+
   #onAppend = null;

   #eventAbortController = null;
@@ -133,7 +135,7 @@ class AnnotationLayerBuilder {
     this.#onAppend?.(div);

     if (annotations.length === 0) {
-      this.hide();
+      this.hide(/* internal = */ true);
       return;
     }

@@ -184,7 +186,9 @@ class AnnotationLayerBuilder {
     this.#eventAbortController = null;
   }

-  hide() {
+  hide(internal = false) {
+    this.#externalHide = !internal;
```

 - Change this code as follows.
```suggestion
    // Don't show the annotationLayer if it was explicitly hidden previously.
    if (!this.#externalHide) {
      this.div.hidden = false;
    }
```

ryzokuken commented:
Thanks for the amazing suggestions, @Snuffleupagus. I just implemented them.

Snuffleupagus commented on web/annotation_layer_builder.js:
```diff
@@ -195,6 +220,45 @@ class AnnotationLayerBuilder {
+
+    if (!this.annotationLayer) {
+      this.#initAnnotationLayer(viewport, structTreeLayer);
+      setLayerDimensions(this.annotationLayer.div, viewport);
```
Can this be shortened?
```suggestion
      setLayerDimensions(this.div, viewport);
```

ryzokuken replied:
Good catch, thanks. Done.

calixteman commented:
@Snuffleupagus anything else before I run the tests ?

Snuffleupagus commented:
> anything else before I run the tests ?

Not off the top of my head, since this is finally starting to look reasonable w.r.t. existing code-paths and edge-cases.

calixteman commented:
/botio test

ryzokuken commented:
At a glance, the integration test failures seem unrelated to this PR, although I'll check to be sure.

calixteman reviewed (approved):
LGTM. Thank you.

calixteman commented:
@Snuffleupagus: gentle ping for your approval.
We'd like to merge it asap in nightly.
In considering your last comment I guess it's fine, but I prefer asking.

Snuffleupagus commented on web/autolinker.js:
```diff
@@ -0,0 +1,147 @@
+  range.setEnd(highlighter.textDivs[end.divIdx].firstChild, end.offset);
+
+  return {
+    id: `added_link_${id}`,
```
```suggestion
    id: `inferred_link_${id}`,
```

Snuffleupagus commented on extensions/chromium/preferences_schema.json:
```diff
@@ -220,6 +220,11 @@
       "default": "CanvasText"
+    },
+    "enableAutoLinking": {
+      "description": "Automatically detect URLs in the text and create links for them",
```
This description should perhaps match the JSDocs for the `PDFViewer`/`PDFPageView` parameter better?

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -1100,10 +1106,18 @@ class PDFPageView {
+        const textLayerPromise = this.#renderTextLayer();
 
         if (this.annotationLayer) {
           await this.#renderAnnotationLayer();
+          if (this.#enableAutoLinking) {
```
```suggestion
          await this.#renderAnnotationLayer();
          
          if (this.#enableAutoLinking) {
```

Snuffleupagus commented on test/unit/autolinker_spec.js:
This test doesn't actually run, since you didn't add an entry for it in https://github.com/mozilla/pdf.js/blob/master/test/unit/jasmine-boot.js

(Sorry to say, but this really should have been obvious during local testing using `gulp unittest` or `gulp test`.)

ryzokuken replied:
done! I was using `unittestcli` to test locally instead and since I'd added the file to that config I didn't notice that I was missing a spot. That said, shouldn't this kind of things be a part of our linting/checking workflow? I don't exactly agree that it's as obvious to a newer contributor as it may be to you.

Snuffleupagus replied:
> I was using `unittestcli` to test locally instead

That's unfortunately not appropriate on its own, since those are Node.js unit-tests which are quite limited in scope and thus won't test everything.

You can still run them, but that should be done *in addition to* the normal browser-based tests such as `gulp test`, `gulp unittest`, `gulp fonttest`, and `gulp integrationtest` (which ones you need depends on the modified code).

> That said, shouldn't this kind of things be a part of our linting/checking workflow?

I don't know if that'd work, since there's three different categories of unit-test files: ones that can run in both browsers *and* Node.js, ones that can only run in browsers, and ones that can only run in Node.js environments.

calixteman commented:
The integration tests failures are unrelated.
@Snuffleupagus gentle ping for the review, thank you.
